### PR TITLE
#90705 Add Filter to Allow Overriding of ICS Single-Event Links

### PR DESCRIPTION
**Pull request by technosailor**
**Ticket:** [**#90705**](http://central.tri.be/issues/90705)

---

Hi Team

This is a small PR that adds a filter for ICS links on single event pages. We could use it on a project and would love to have it in the next MR if possible.

### CODE REVIEW

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Expected next thing to be an escaping function (see Codex for 'Data Validation'), not ')' | [iCal.php:83](https://github.com/moderntribe/the-events-calendar/blob/4df7aa3c1e5884059de8cbf1177119cd910b9962/src/Tribe/iCal.php#L83)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Expected next thing to be an escaping function (see Codex for 'Data Validation'), not ')' | [iCal.php:83](https://github.com/moderntribe/the-events-calendar/blob/8646dc4ba661bb9bf6a05e8ac8c7924bf46b9d3d/src/Tribe/iCal.php#L83)

> via jenkins/codesniffer-phpcs

File: src/Tribe/iCal.php [hunk 1]
```diff
@@ -75,12 +75,15 @@ public function single_event_links() {
...
-		echo '<a class="tribe-events-ical tribe-events-button" href="' . esc_url( tribe_get_single_ical_link() ) . '" title="' . esc_attr__( 'Download .ics file', 'the-events-calendar' ) . '" >+ ' . esc_html__( 'iCal Export', 'the-events-calendar' ) . '</a>';
-		echo '</div><!-- .tribe-events-cal-links -->';
+		$calendar_links = '<div class="tribe-events-cal-links">';
+		$calendar_links .= '<a class="tribe-events-gcal tribe-events-button" href="' . Tribe__Events__Main::instance()->esc_gcal_url( tribe_get_gcal_link() ) . '" title="' . esc_attr__( 'Add to Google Calendar', 'the-events-calendar' ) . '">+ ' . esc_html__( 'Google Calendar', 'the-events-calendar' ) . '</a>';
+		$calendar_links .= '<a class="tribe-events-ical tribe-events-button" href="' . esc_url( tribe_get_single_ical_link() ) . '" title="' . esc_attr__( 'Download .ics file', 'the-events-calendar' ) . '" >+ ' . esc_html__( 'iCal Export', 'the-events-calendar' ) . '</a>';
+		$calendar_links .= '</div><!-- .tribe-events-cal-links -->';
+
+		echo apply_filters( 'tribe_events_single_links', esc_html( $calendar_links ) );
```
Comment (barryhughes) [inline src/Tribe/iCal.php]: The text contains HTML, so we probably want to drop the use of `esc_html()` here :-)

Comment (ggwicz) [reply src/Tribe/iCal.php]: @barryhughes would you think something like `wp_kses_post()` is unnecessary and expensive, or would that be cool to use instead?

Comment (barryhughes) [reply src/Tribe/iCal.php]: Not completely closed to the idea, but I don't feel it's necessary/it's not something we generally do in this sort of situation.

Comment (ggwicz) [reply src/Tribe/iCal.php]: Right on.

**Comment by ggwicz** [comment]
Setting to `hold` for now — I just need to make a handful of changes that are noted in note #1 here: https://central.tri.be/issues/90705#note-1

Will make those changes and set this back to needing a `code review`.

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Expected next thing to be an escaping function (see Codex for 'Data Validation'), not ')' | [iCal.php:88](https://github.com/moderntribe/the-events-calendar/blob/4a0543ed4edec51040fd8e2b59b39066361c36cb/src/Tribe/iCal.php#L88)
:x: | Whitespace found at end of line | [iCal.php:358](https://github.com/moderntribe/the-events-calendar/blob/4a0543ed4edec51040fd8e2b59b39066361c36cb/src/Tribe/iCal.php#L358)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Expected next thing to be an escaping function (see Codex for 'Data Validation'), not ')' | [iCal.php:88](https://github.com/moderntribe/the-events-calendar/blob/48637e4198eec044f7eecfe7f9cf58e14466f5de/src/Tribe/iCal.php#L88)
:x: | Whitespace found at end of line | [iCal.php:358](https://github.com/moderntribe/the-events-calendar/blob/48637e4198eec044f7eecfe7f9cf58e14466f5de/src/Tribe/iCal.php#L358)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Expected next thing to be an escaping function (see Codex for 'Data Validation'), not ')' | [iCal.php:88](https://github.com/moderntribe/the-events-calendar/blob/69e34a863af109969dc384922b8b076e58cff138/src/Tribe/iCal.php#L88)
:x: | Whitespace found at end of line | [iCal.php:358](https://github.com/moderntribe/the-events-calendar/blob/69e34a863af109969dc384922b8b076e58cff138/src/Tribe/iCal.php#L358)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Expected next thing to be an escaping function (see Codex for 'Data Validation'), not ')' | [iCal.php:88](https://github.com/moderntribe/the-events-calendar/blob/b32656d57aa94190cf1c5495ae03ffedae326f4b/src/Tribe/iCal.php#L88)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Expected next thing to be an escaping function (see Codex for 'Data Validation'), not ')' | [iCal.php:88](https://github.com/moderntribe/the-events-calendar/blob/c78f5ba9c19a52202d97ca52ab41047588b3731d/src/Tribe/iCal.php#L88)

> via jenkins/codesniffer-phpcs

File: src/Tribe/iCal.php [hunk 2]
```diff
@@ -75,37 +75,58 @@ public function single_event_links() {
...
+		$calendar_links .= '<a class="tribe-events-gcal tribe-events-button" href="' . Tribe__Events__Main::instance()->esc_gcal_url( tribe_get_gcal_link() ) . '" title="' . esc_attr__( 'Add to Google Calendar', 'the-events-calendar' ) . '">+ ' . esc_html__( 'Google Calendar', 'the-events-calendar' ) . '</a>';
+		$calendar_links .= '<a class="tribe-events-ical tribe-events-button" href="' . esc_url( tribe_get_single_ical_link() ) . '" title="' . esc_attr__( 'Download .ics file', 'the-events-calendar' ) . '" >+ ' . esc_html__( 'iCal Export', 'the-events-calendar' ) . '</a>';
+		$calendar_links .= '</div><!-- .tribe-events-cal-links -->';
+
+		/**
+		 * Allow for complete customization of the iCal and gCal single-event links.
+		 *
+		 * @param string $calendar_links The HTML of the iCal and gCal single-event link buttons.
+		 */
+		echo apply_filters( 'tribe_events_single_links', $calendar_links );
```
Comment (barryhughes) [inline src/Tribe/iCal.php]: Not super stoked on the name of the hook - `tribe_events_single_links` is really generic and doesn't convey anything about it being the iCal/GCal link.

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Expected next thing to be an escaping function (see Codex for 'Data Validation'), not ')' | [iCal.php:88](https://github.com/moderntribe/the-events-calendar/blob/286e8854f82f75c9b631be962fdafa1990e3bd9b/src/Tribe/iCal.php#L88)

> via jenkins/codesniffer-phpcs

**Comment by ggwicz** [comment]
Good call @barryhughes — I've changed the filter name to `tribe_events_ical_single_event_links`, so this one's ready for another review 🔍

**Pull request merged**